src/app/decorations/qgsdecorationimage.cpp

```diff
@@ -0,0 +1,279 @@
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsdecorationimage.h"
+
+#include "qgsdecorationimagedialog.h"
+
+#include "qgisapp.h"
+#include "qgscoordinatetransform.h"
+#include "qgsexception.h"
+#include "qgsimagecache.h"
+#include "qgslogger.h"
+#include "qgsmaplayer.h"
+#include "qgsproject.h"
+#include "qgssymbollayerutils.h"
+#include "qgssvgcache.h"
+
+#include <QPainter>
+#include <QMenu>
+#include <QDir>
+#include <QFile>
+#include <QSvgRenderer>
+
+#include <cmath>
+#include <cassert>
+
+
+QgsDecorationImage::QgsDecorationImage( QObject *parent )
+  : QgsDecorationItem( parent )
+{
+  mPlacement = BottomLeft;
+  mMarginUnit = QgsUnitTypes::RenderMillimeters;
+
+  setName( "Image" );
+  projectRead();
+}
+
+void QgsDecorationImage::projectRead()
+{
+  QgsDecorationItem::projectRead();
+  mColor = QgsSymbolLayerUtils::decodeColor( QgsProject::instance()->readEntry( mNameConfig, QStringLiteral( "/Color" ), QStringLiteral( "#000000" ) ) );
+  mOutlineColor = QgsSymbolLayerUtils::decodeColor( QgsProject::instance()->readEntry( mNameConfig, QStringLiteral( "/OutlineColor" ), QStringLiteral( "#FFFFFF" ) ) );
+  mSize = QgsProject::instance()->readDoubleEntry( mNameConfig, QStringLiteral( "/Size" ), 16.0 );
+  setImagePath( QgsProject::instance()->readEntry( mNameConfig, QStringLiteral( "/ImagePath" ), QString() ) );
+  mMarginHorizontal = QgsProject::instance()->readNumEntry( mNameConfig, QStringLiteral( "/MarginH" ), 0 );
+  mMarginVertical = QgsProject::instance()->readNumEntry( mNameConfig, QStringLiteral( "/MarginV" ), 0 );
+}
+
+void QgsDecorationImage::saveToProject()
+{
+  QgsDecorationItem::saveToProject();
+  QgsProject::instance()->writeEntry( mNameConfig, QStringLiteral( "/Color" ), QgsSymbolLayerUtils::encodeColor( mColor ) );
+  QgsProject::instance()->writeEntry( mNameConfig, QStringLiteral( "/OutlineColor" ), QgsSymbolLayerUtils::encodeColor( mOutlineColor ) );
+  QgsProject::instance()->writeEntry( mNameConfig, QStringLiteral( "/Size" ), mSize );
+  QgsProject::instance()->writeEntry( mNameConfig, QStringLiteral( "/ImagePath" ), mImagePath );
+  QgsProject::instance()->writeEntry( mNameConfig, QStringLiteral( "/MarginH" ), mMarginHorizontal );
+  QgsProject::instance()->writeEntry( mNameConfig, QStringLiteral( "/MarginV" ), mMarginVertical );
+}
+
+// Slot called when the buffer menu item is activated
+void QgsDecorationImage::run()
+{
+  QgsDecorationImageDialog dlg( *this, QgisApp::instance() );
+  dlg.exec();
+}
+
+void QgsDecorationImage::setImagePath( const QString &imagePath )
+{
+  mImagePath = imagePath;
+  mImageFormat = FormatUnknown;
+
+  QString resolvedPath = QgsSymbolLayerUtils::svgSymbolNameToPath( mImagePath, QgsProject::instance()->pathResolver() );
+  QFileInfo fileInfo( resolvedPath );
+  if ( fileInfo.exists() )
```

Comment:
```suggestion
  if ( QFileInfo::exists( resolvedPath ) )
```

(it's faster)

Reply:
You suggest using that faster path even if we ultimately need a QFileInfo object to get the file suffix?

Reply:
Missed that :P ignore me!